lens/history.go

```diff
@@ -0,0 +1,279 @@
+// licenses/APL.txt.
+
+package lens
+
+import (
+	"context"
+
+	"github.com/ipfs/go-datastore/query"
+	"github.com/sourcenetwork/immutable"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/core"
+	"github.com/sourcenetwork/defradb/datastore"
+)
+
+// schemaHistoryLink represents an item in a particular schema's history, it
+// links to the previous and next version items if they exist.
+type schemaHistoryLink struct {
+	// The schema version id of this history item.
+	schemaVersionID string
+
+	// The history link to the next schema version, if there is one
+	// (for the most recent schema version this will be None).
+	next immutable.Option[*schemaHistoryLink]
+
+	// The history link to the previous schema version, if there is
+	// one (for the initial schema version this will be None).
+	previous immutable.Option[*schemaHistoryLink]
+}
+
+// targetedSchemaHistoryLink represents an item in a particular schema's history, it
+// links to the previous and next version items if they exist.
+//
+// It also contains a vector which describes the distance and direction to the
+// target schema version (given as an input param on construction).
+type targetedSchemaHistoryLink struct {
+	// The schema version id of this history item.
+	schemaVersionID string
+
+	// The link to next schema version, if there is one
+	// (for the most recent schema version this will be None).
+	next immutable.Option[*targetedSchemaHistoryLink]
+
+	// The link to the previous schema version, if there is
+	// one (for the initial schema version this will be None).
+	previous immutable.Option[*targetedSchemaHistoryLink]
+
+	// The distance and direction from this history item to the target.
+	//
+	// A zero value indicates that this is the target item. A positive value
+	// indicates that the target is more recent. A negative value indicates
+	// that the target predates this history item.
+	targetVector int
+}
+
+// getTargetedSchemaHistory returns the history of the schema of the given id, relative
+// to the given target schema version id.
+//
+// This includes any history items that are only known via registered
+// schema migrations.
+func getTargetedSchemaHistory(
+	ctx context.Context,
+	txn datastore.Txn,
+	lensConfigs []client.LensConfig,
+	schemaID string,
+	targetSchemaVersionID string,
+) (map[schemaVersionID]*targetedSchemaHistoryLink, error) {
+	history, err := getSchemaHistory(ctx, txn, lensConfigs, schemaID)
+	if err != nil {
+		return nil, err
+	}
+
+	result := map[schemaVersionID]*targetedSchemaHistoryLink{}
+
+	for _, item := range history {
+		result[item.schemaVersionID] = &targetedSchemaHistoryLink{
+			schemaVersionID: item.schemaVersionID,
+		}
+	}
+
+	for _, item := range result {
+		schemaHistoryLink := history[item.schemaVersionID]
+		nextHistoryItem := schemaHistoryLink.next
+		if !nextHistoryItem.HasValue() {
+			continue
+		}
+		nextItem := result[nextHistoryItem.Value().schemaVersionID]
```

Comment:
question: isn't `nextHistoryItem.Value()` a pointer? What makes you sure it's not `nil`?